WIP: Add LXD driver & tests

pockerman opened the pull request:
##WIP: Add support for LXC containers

### Description

Add support for LXC containers.
- Add ```container/drivers/lxd.py``` module with the following classes
- ```LXDAPIException``` simple exception class
-  ```LXDStoragePool``` wrapper of the info of a storage pool
- ```LXDServerInfo``` wrapper of the info of the LXC server
- ```LXDResponse``` represents the response to an LXD API call
- ```LXDConnection``` represents a connection
- ```LXDtlsConnection``` TLS connection
- ```LXDContainerDriver``` main class wrapping the LXD API

### Status

- work in progress

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

pockerman committed:
Add LXD driver & tests

Kami committed:
Update setup.py and drop support for Python 2.7 and Python 3 < 3.4.0.

Kami committed:
Update tox and travis config and drop support for Python 2.7 and Python
3 < 3.4.0.

Also move all the various check targets to use Python 3.7.

Kami committed:
Use working version.

Kami committed:
Also drop Pythohn 3.4.

Kami committed:
Update setup.py.

Kami committed:
Fix lint issue which arose now that we run lint targets under Python 3
and make sure all the examples are Python 3 compatiblr.

Kami committed:
Add missing entry to tox config.

Kami committed:
Use pyp3.5.

Kami committed:
Update pylint and astroud, update plugin so it works with newer version
of astroid.

Kami committed:
Fix various false positives and actual lint issues detected by new
version of pylint.

Kami committed:
Update tox config.

Kami committed:
Fix pylint issue.

Kami committed:
Use latest version of coverage.

Kami committed:
Fix failing test.

Kami committed:
Require Python 3.5.0.

Kami committed:
Merge remote-tracking branch 'remotes/upstream/trunk' into drop_support_for_old_python_versions

Kami committed:
Merge branch 'trunk' of http://gitbox.apache.org/repos/asf/libcloud into drop_support_for_old_python_versions

Kami committed:
Add changelog and upgrade notes entry.

Kami committed:
Remove vSphere driver which doesn't support Python 3 and document that
in upgrade notes, changelog and documentation.

Kami committed:
Remove dependencies which are not needed anymore.

Kami committed:
Re-generate support provider tables.

Kami committed:
Add note on supported Python versions to readme.

Kami committed:
Fix bad merge resolution, remove attributes which are not needed.

Kami committed:
Use same version of setuptools everywhere.

Kami committed:
Use latest version of test dependencies.

Kami committed:
Simplify travis config.

Kami committed:
Also run tests with PyPy 3.6.

Kami committed:
Use user friendly travis job names.

Kami committed:
PyPy 3.6 is not supported by travis yet.

Kami committed:
Add documentation section on supported Python versions.

Kami commented:
Thanks for the contribution.

Since this is a larger contribution, can you please also work on signing an ICLA (https://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes)?

pockerman commented:
yes of course

Kami committed:
Bump version to v2.8.0.

Kami committed:
Update available ec2 instance types.

Kami committed:
Indicate we are now working on 2.9.0-dev.

Kami committed:
Update "ex_list_instancegroups" method in the GCE driver so it doesn't
throw an exception if a response doesn't contain "zone" attribute.

Reported by Kartik Subbarao.

Resolves #1346

Kami committed:
Add changelog entry.

Kami committed:
Fix changelog formatting.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+    return host
+
+
+def check_certificates(key_file, cert_file, **kwargs):
```
It appears that some of that logic is similar to the one in #1394 so once those PRs are merged, both of the drivers could perhaps be refactored to use those common utility functions.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+    # if the type of the response is an error
+    if response_dict['type'] == LXD_ERROR_STATUS_RESP:
+        # an error returned
+        raise LXDAPIException(message="response type is error")
```
For ease of debugging / troubleshooting, would it make sense to also include ``respose_dict`` value in the actual exception error message?

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+            if content_type == 'application/json' or content_type == '':
+                if self.headers.get('transfer-encoding') == 'chunked' and \
+                        'fromImage' in self.request.url:
+                    body = [json.loads(chunk) for chunk in
```
Eventually it would be nice if we could refactor some of the transfer-encoding chunked logic to a base ``ChunkedEncodingJsonResponse`` or similar Response class.

pockerman replied:
Basically, this is a copy verbatim from the Docker driver which means I should work more on it at some point. PyCharm shows 138 implementations of ```JsonResponse``` we need someone brave :) :)

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                error_msg = m.group(1)
+                raise Exception(error_msg)
+            else:
+                raise Exception(
```
Should we also include original response in the error message body for ease of debugging?

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                 host='localhost', port=8443, ca_cert='',
+                 key_file=None, cert_file=None, **kwargs):
+
+        if 'certificate_validator' in kwargs.keys():
```
Please explicitly declare supported arguments (``certificate_validator``) which are known up front instead of using ``**kwargs`` (I know a lot of the existing drivers did, but we just tried to clean and improve some of that legacy code).

pockerman replied:
basically I am seriously thinking to remove this altogether...I put it there initially to get more informative errors when trying to use TLS. I am happy to so as you suggested but what do you think about removing it completely?

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+        meta = response_dict["metadata"]
+        return LXDServerInfo.build_from_response(metadata=meta)
+
+    def deploy_container(self, name, image, cluster=None,
```
Please prefix arguments which are not part of the standard container API with ``ex_`` (profiles, ephemeral, config, etc,).

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+    def deploy_container(self, name, image, cluster=None,
+                         parameters=None, start=True,
+                         architecture=default_architecture,
+                         profiles=[default_profiles],
```
You don't appear to be mutating ``profiles`` inside the function body, but it would still be safer and less error prone to avoid mutable defaults (https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) and something like ``profiles = None``
and inside the body ``profiles = proxies or [default_profiles]``).

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+
+        return container
+
+    def get_container(self, id, get_ip_addr=True):
```
Please prefix ``get_ip_addr`` with ``ex_`` since it's not part of the standard API.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+        metadata.update({"ips": ips})
+        return self._to_container(metadata=metadata)
+
+    def start_container(self, container, timeout=default_time_out):
```
Please prefix ``timeout`` with ``ex_`` since it's not part of the standard API.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                                         timeout=timeout,
+                                         force=True, stateful=True)
+
+    def stop_container(self, container, timeout=default_time_out):
```
Please prefix ``timeout`` with ``ex_`` since it's not part of the standard API.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                                         timeout=timeout,
+                                         force=True, stateful=True)
+
+    def restart_container(self, container, timeout=default_time_out):
```
Please prefix ``timeout`` with ``ex_`` since it's not part of the standard API.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                                         timeout=timeout,
+                                         force=True, stateful=True)
+
+    def destroy_container(self, container, timeout=default_time_out):
```
Please prefix ``timeout`` with ``ex_`` since it's not part of the standard API.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+
+        # if the container is running then we cannot delete
+
+        if container.state == ContainerState.RUNNING:
```
I'm fine with doing that check here although we don't really have clear rules / guidelines on how to handle such scenarios in Libcloud drivers.

A lot of drivers simply let the API handle that validation instead of doing it in the driver. This way it still works if user doesn't have the latest reference to the ``container`` object locally.

So another option would be to simply remove that check here.

pockerman replied:
ok I can remove it then...

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+                              extra=None)
+        return container
+
+    def list_containers(self, image=None, cluster=None, detailed=True):
```
Please prefix ``detailed`` with ``ex_``.

Kami commented on libcloud/container/drivers/lxd.py:
```diff
@@ -0,0 +1,1269 @@
+
+        return self._to_image(metadata=response_dict['metadata'])
+
+    def install_image(self, path, **description_dict):
```
Ideally, all those non standard arguments would be prefixed with ``ex_`` or you would take  a single ``ex_description_dict`` argument which is a dictionary.

Kami commented:
Nice work and thanks again for the contribution.

I saw some similarities with #1394, are those PRs somewhat related?

Kami committed:
Merge branch 'trunk' into drop_support_for_old_python_versions

Kami committed:
Signal those changes will represent v3.0.0 release.

Kami committed:
Merge pull request #1377 from Kami/drop_support_for_old_python_versions

Drop support for Python 2.7 and Python 3 < 3.5.0

Kami committed:
Use latest version of pip on travis.

Kami committed:
Try also running unit tests on windows.

Kami committed:
Use pip3.

Kami committed:
Use python 3.7, set PATH.

Kami committed:
Use latest Python 3.7 version.

Kami committed:
Specify correct tox env.

Kami committed:
Use special tox target for Windows unit tests which doesn't rely on
libvirt.

Kami committed:
Ignore test failure due to permission error in Windows.

Kami committed:
Make tests OS agnostic.

Kami committed:
Make sure we clean up in tearDown.

Kami committed:
Skip this test on Windows.

Kami committed:
That method is now also supported on Windows in newer versions of
Python.

Kami committed:
Use a temporary file which exists.

Kami committed:
Fix the OS check.

Kami committed:
Use new / non-deprecated test method names.

Kami committed:
Add changelog entry.

Kami committed:
Merge pull request #1396 from Kami/run_unit_tests_on_multiple_oses

Also run unit tests on Windows using Python 3.7

pockerman commented:
both contributions are part of an extension underway for mist.io

pockerman committed:
Fix some review findings. Small improvements on the implemented API

yukw777 committed:
Support for aws ec2 spot instances

pockerman committed:
Merge from mist lxc fork branch

Kami committed:
Add info about v2.8.0 release.

pockerman committed:
Fix flake8 issues. Add freeze/unfreeze constainer state methods

bit-project-de committed:
fix curl command

1. Header must use "X-Ovh-Application" to correctly provide the App Key for OVH Driver.
2. Comma at end of accessRules Array causes "invalid JSON" response.

Kami committed:
Merge pull request #1402 from bit-project-de/patch-1

fix curl command

Kami committed:
Merge pull request #1398 from yukw777/ec2-spot

Support for aws ec2 spot instances

Kami committed:
Add changelog entry.

Kami committed:
Merge branch 'lxd_driver' of https://github.com/pockerman/libcloud into pockerman-lxd_driver

Kami committed:
Add missing license header.

Kami committed:
Fix lint (add missing required argument).

Kami commented:
Thanks for addressing the review comments.

There are still a couple of things which would need some work and improvements (refactoring), but those are not blockers so we can go ahead and merge this as is and improve other things in separate PRs.

(I also pushed two small changes to fix the build)

Kami committed:
Include original response body in the exception error message to make
troubleshooting easier.

Kami committed:
Fix changelog formatting, add entry for changes in #1395.
